### DISCUSSION TITLE
getting error 401 when calling issue_documents via axios, but it works when calling via Postman

### DISCUSSION THREAD

**Discussion by pgururani** [fattureincloud/api#242] [Q&A] [answered]
hi, we're migrating to fattureincloud V2, and facing the issue with error code **401**, in fatture-in-cloud we have setup application and connected it to the company with the following permissions

![image](https://user-images.githubusercontent.com/58684557/214319350-04a2bfe0-ff11-4fa3-864b-328fd7acf11c.png)

below is my use case scenario

App visibility - **private app**
usage - we're using Manual authentication
url: https://api-v2.fattureincloud.it/c/{{company_id}}/issued_documents
authentication: Bearer token (access token generated in the application) 

My Axios call for generating an invoice is giving me an error "Request failed with status code 401"
My company id and token seems fine, since i am able to generate invoice through postman

below is the sample request

{"data": {
  "type":"invoice",
  "entity":{
    "id":1,
    "name":"test user",
    "vat_number":"12345678910",
    "tax_code":"1234567",
    "address_street":"My Address",
    "address_postal_code":"13500",
    "address_city":"My City",
    "address_province":"AL",
    "country":"Italia"
  },
  "date":"2023-01-24",
  "subject":"test",
  "visible_subject":"test",
  "currency":{
    "id":"EUR",
    "exchange_rate":"1.00000",
    "symbol":"€"
  },
  "language":{
    "code":"it",
    "name":"Italiano"
  },
  "e_invoice":true,
  "ei_data":{
    "payment_method":"MP08"
  },
  "use_gross_prices":true,
  "items_list":[
    {
      "apply_withholding_tax":false,
      "name":"item 1",
      "description":"Abbonamento regalo",
      "gross_price":"19.99",
      "category":"cucina",
      "discount":0,
      "qty":1,
      "vat":{
        "id":4
      }
    }
  ],
  "payments_list":[
    {
      "amount":"19.99",
      "due_date":"2023-01-24",
      "status":"not_paid"
    }
  ],
  "use_split_payment":false,
  "options":{
    "fix_payments":true
  }
}}

[axios call sample]

axios({
          method: 'post',
          url: url,
          headers: {
            Authorization: 'Bearer ACCESS_TOKEN',
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          data
        }).then(response => {
          if (_.has(response, ['data', 'error']) && response.data.error !== 'undefined') {
            console.log(`E-Invoice generation failed for email ${userDetails.email}`, JSON.stringify(response.data));
          } else {
            console.log(`E-Invoice generated successfully for email ${userDetails.email}`, JSON.stringify(response.data));
          }
          return contextData;
        }).catch(err => {
          console.log(`error at generate-user-invoice hook for email ${userDetails.email} for error ${JSON.stringify(err)}`);
          return contextData;
        });

please let me know the possible reasons
thanks

**Comment by valmoz** [comment, marked as answer]
Hi @pradeep-gururani-incaendo 
**first of all I edited your message to remove the Access Token from the code.**
It is sensible info and it should **never** be posted on a public page like this one.

I would strongly suggest you cancel it and recreate it from scratch because we must consider it unsafe right now.

Mauro

**Reply by pgururani** [reply]
hi, thanks I appreciate it, actually the token I added is not real, its having some made-up random string

can you please suggest some possible solutions for the above issue ?

**Reply by valmoz** [reply]
From what you posted here everything seems ok.

Just be sure to insert the company ID and the token correctly into the request.

I suggest you write to our customer support, so we can access the logs to investigate this issue. You can attach the link to this discussion in the support request to speed up the process.

**Reply by pgururani** [reply]
@valmoz thanks, the companyID was added correctly and it was working fine via postman, but it fails when we do the same thing via api (nodejs  backend), will making the app's visibility **public** help in such scenario ?


I dropped them a message via the chat feature in **fattureincloud** dashboard, I'll attach the link too
<img width="447" alt="image" src="https://user-images.githubusercontent.com/58684557/214355911-221fd1aa-af65-4781-90ec-d24f49529a99.png">

**Reply by valmoz** [reply]
The public app is useful only if you need to manage multiple FIC accounts you don't know right now, for example because you're developing an e-commerce integration that you want to sell to external users. Otherwise the private app is more than enough.

you can check the [App Visibility](https://developers.fattureincloud.it/docs/basics/create-an-app#-app-visibility) page for more info.

BTW, if the request works with Postman but not with Axios, most probably the problem is that some parameter was not set up correctly in the Axios request. We'll have more info once we'll access the logs.

Thanks
Mauro

**Reply by pgururani** [reply]
Thanks @valmoz I'll wait for the logs report 👍

**Reply by pgururani** [reply]
hey @valmoz did you get a chance to have a look in the logs

**Reply by valmoz** [reply]
Hi @pradeep-gururani-incaendo we need for the customer support team to take charge of the request. 
Usually, it will require at most 24 hours, but it will probably be faster than that.

**Reply by pgururani** [reply]
thanks 👍

**Reply by valmoz** [reply]
Hi @pradeep-gururani-incaendo, I was able to check the logs for your company id. I can only see the logs for the successful requests from Postman, but I'm unable to see the 401 requests from Axios.

This usually happens when the access token or the company id (or possibly both) are not inserted correctly into the request.
I suggest you try to check again if the access token and the company id are inserted correctly into the Axios request.

I would also suggest trying first to perform from Axios a [List User Companies](https://developers.fattureincloud.it/api-reference#get-/user/companies) request (since it doesn't require the company ID) and if it works to also try a [List Issued Documents](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents) (since it doesn't have a body) to check if the request returns a success or not.

Let us know.

Thanks
Mauro

**Reply by pgururani** [reply]
ok, thanks, I'll check

**Reply by valmoz** [reply]
@pradeep-gururani-incaendo 
Since it works correctly from Postman, I also suggest you try to generate the code automatically from postman using this guide: [Vanilla Code Examples](https://developers.fattureincloud.it/docs/quickstarts/vanilla-code).
You can try to compare your code with the one generated by Postman and see the differences.

Thanks 
Mauro

**Reply by pgururani** [reply]
@valmoz  hey, I checked the **list issued documents** via **xhr request** and it sent the response fine, can there be issue with the body of the "create issue document" request

below is the code used via xhr

const XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;

const xhr = new XMLHttpRequest();

xhr.open( 'GET', 'https://api-v2.fattureincloud.it/c/compID/issued_documents/?type=invoice');

xhr.setRequestHeader('Accept', 'application/json');
xhr.setRequestHeader('Authorization', 'Bearer AUTH_TOKEN');

try {
  xhr.onload = function() {
    console.log('xhr-status', JSON.stringify(xhr, null, 2));
    console.log('xhr-status', xhr.status);
    if ( xhr.status == 200 ) {
      console.log( xhr.responseText );
    } else {
      console.log( `Error: ${xhr.status}` );
    }
  };
} catch (error) {
  console.log('error in req', error);
}

xhr.send();

**Reply by emajo** [reply]
Hi @pradeep-gururani-incaendo, so you are still having issues creating a document? If so can you share what kind of error you are facing?
If you haven't already considered to use our [js-sdk](https://www.npmjs.com) I strongly suggest to do so as it simplifies a lot making the requests.

Thanks,
Emanuele

**Reply by pgururani** [reply]
@hey @emajo I tried the approach suggested by @valmoz and got a `success` response converting `POSTMAN` to `XHR code snippet` and sending the request via xhr

Thanks for your help guys, I made some silly mistake in the code, but it works fine now
